scripts: size_report: Add sorting key and depth as cmake arguments

adleris opened the pull request:
This change brings the ability to specify the sort key (by name or by size) and report depth as arguments to cmake/west. They can be given like:

`west build -t rom_report -- -Dreport_depth=4 -Dsort_key=size`

Addresses #57569

adleris committed:
scripts: size_report: Add sorting key and depth as cmake arguments

This change brings the ability to specify the sort key (by name or
by size) and report depth as arguments to cmake/west. They can be
given like:

west build -t rom_report -- -Dreport_depth=4

Addresses #57569

Signed-off-by: Alexander Leris <alexander.leris@gmail.com>

nordicjm commented on cmake/reports/CMakeLists.txt:
```diff
@@ -3,7 +3,11 @@
 set(flag_for_footprint all -q)
-set(report_depth 99)
+
+set(report_depth 99 CACHE STRING "Depth of tree to display")
+set(sort_key "name" CACHE STRING "Sort key for results display (name or size)")
+set_property(CACHE sort_key PROPERTY STRINGS "name" "size")
```
So a full rebuild is required to change these? I cannot accept that limitation.

adleris replied:
Firstly, thank you both for the review.

I must admit I am somewhat of a novice at using CMake, however from my testing of the script it seemed like this was the appropriate way to pass the parameters through to the script. As @tejlmand said, it is a much shorter build that a full rebuild.

I hadn't considered Kconfig, but it seems like this would be an option that you might want to change without the full recompile. However, I'd be happy to defer to both of your expertise on the best method to implement this.

tejlmand replied:
I would normally prefer Kconfig, as that is mostly used, but exactly for the concern raised by @nordicjm, and which I share, then I think a CMake variable is the best choice in this case.

would like to wait for @nordicjm to see if his concerns has been addressed.

tejlmand commented on cmake/reports/CMakeLists.txt:
```diff
@@ -40,6 +44,7 @@ if (CONFIG_BUILD_WITH_TFM)
       -o ${CMAKE_BINARY_DIR}
       ${workspace_arg}
       -d ${report_depth}
+      -s ${sort_key}
```
why is this only added for the TF-M reports, and not the general Zephyr report ?

adleris replied:
This was an error on my part, I've addressed this in `8f2ac38`

adleris commented on cmake/reports/CMakeLists.txt:
```diff
@@ -3,7 +3,11 @@
 set(flag_for_footprint all -q)
-set(report_depth 99)
+
+set(report_depth 99 CACHE STRING "Depth of tree to display")
+set(sort_key "name" CACHE STRING "Sort key for results display (name or size)")
```
I'd agree with that.

Perhaps names such as `FOOTPRINT_REPORT_DEPTH` and `FOOTPRINT_REPORT_SORT_KEY`?

tejlmand replied:
> Perhaps names such as `FOOTPRINT_REPORT_DEPTH` and `FOOTPRINT_REPORT_SORT_KEY`?

Sounds fine with me.
@nordicjm any thoughts ?

adleris committed:
scripts: size_report: Add sort flag to other report options

The sort flag was only added to the TF-M reports, this change
brings it to the general Zephyr build and to TF-M BL2 builds.

Signed-off-by: Alexander Leris <alexander.leris@gmail.com>

tejlmand commented on cmake/reports/CMakeLists.txt:
```diff
@@ -3,7 +3,11 @@
+set(report_depth 99 CACHE STRING "Depth of tree to display")
+set(sort_key "name" CACHE STRING "Sort key for results display (name or size)")
+set_property(CACHE sort_key PROPERTY STRINGS "name" "size")
+
```
please use a `zephyr_get(FOOTPRINT_REPORT_DEPTH)` for this variable and related ones. (if of course CMake variables are agreed to be used.)

This ensures a uniform way of getting CMake variables in Zephyr, as well as include support for variables passed through sysbuild .

See more here: https://github.com/zephyrproject-rtos/zephyr/blob/8d6b206064c915316245616a718ae29eb157a717/cmake/modules/extensions.cmake#L2469-L2497

tejlmand commented:
@nordicjm any replies to the comments where your feedback would be appreciated  ?
Ref: 
- https://github.com/zephyrproject-rtos/zephyr/pull/57633#discussion_r1190869007
- https://github.com/zephyrproject-rtos/zephyr/pull/57633#discussion_r1190866394
